File: javascript/gift_certificate/getGiftCertificatesByQuery.js
import { giftCertificateApi } from '../api.js';

// ReSharper disable once ClassNeverInstantiated.Global
export class GetGiftCertificatesByQuery {
    static async execute() {
        const giftCertificates = await this.getGiftCertificateByQueryCall();
        if(giftCertificates !== undefined) {
            for (const giftCertificate of giftCertificates) {
                // Assuming Utility is available in your TS environment
                console.log(giftCertificate);
            }
        }
    }

    static async getGiftCertificateChunk(api, offset, limit) {
        const expansion = "ledger";

        // leaving query empty, so no filtering, and I should get all records returned.
        const query = {};

        const gcResponse = await new Promise((resolve, reject) => {
            api.getGiftCertificatesByQuery(
                query, {_limit: limit, _offset: offset, _expand: expansion},
                function(error, data, response) {
                    if (error) {
                        reject(error);
                    } else {
                        resolve(data);
                    }
                }
            );
        });

        if (gcResponse !== undefined && gcResponse.success === true && gcResponse.gift_certificates !== null) {
            return gcResponse.gift_certificates;
        }

        return [];
    }

    // ReSharper disable once MemberCanBePrivate.Global
    static async getGiftCertificateByQueryCall() {
        const api = giftCertificateApi;

        const giftCertificates = [];

        let iteration = 1;
        let offset = 0;
        const limit = 200;
        let moreRecordsToFetch = true;

        while (moreRecordsToFetch) {
            console.log(`executing iteration ${iteration}`);
            const chunkOfCertificates = await this.getGiftCertificateChunk(api, offset, limit);
            if(chunkOfCertificates !== undefined && chunkOfCertificates !== null) {
                giftCertificates.push(...chunkOfCertificates);
                offset += limit;
                moreRecordsToFetch = chunkOfCertificates.length === limit;
                iteration++;
            } else {
                moreRecordsToFetch = false;
            }
        }

        return giftCertificates;
    }
}

File: csharp/gift_certificate/GetGiftCertificatesByQuery.cs
using System.Collections.Generic;
using com.ultracart.admin.v2.Api;
using com.ultracart.admin.v2.Model;

namespace SdkSample.gift_certificate
{
    // ReSharper disable once ClassNeverInstantiated.Global
    public class GetGiftCertificatesByQuery
    {

        public static void Execute()
        {
            var giftCertificates = GetGiftCertificateByQueryCall();
            foreach (var giftCertificate in giftCertificates)
            {
                Utility.DumpObject(giftCertificate, "Gift Certificate");    
            }
        }


        private static List<GiftCertificate> GetGiftCertificateChunk(GiftCertificateApi api, int offset, int limit)
        {
            const string expansion = "ledger";

            // leaving query empty, so no filtering, and I should get all records returned.
            GiftCertificateQuery query = new GiftCertificateQuery();
            
            var gcResponse = api.GetGiftCertificatesByQuery(query, limit, offset, null, null, expansion);
                if(gcResponse.Success == true && gcResponse.GiftCertificates != null){
                    return gcResponse.GiftCertificates;
                }

                return new List<GiftCertificate>();
        }
        
        
        
        // ReSharper disable once MemberCanBePrivate.Global
        public static List<GiftCertificate> GetGiftCertificateByQueryCall()
        {
            var api = new GiftCertificateApi(Constants.ApiKey);

            List<GiftCertificate> giftCertificates = new List<GiftCertificate>();

            var iteration = 1;
            var offset = 0;
            var limit = 200;
            var moreRecordsToFetch = true;

            while( moreRecordsToFetch ){

                System.Console.WriteLine("executing iteration " + iteration);
                var chuckOfCertificates = GetGiftCertificateChunk(api, offset, limit);
                giftCertificates.AddRange(chuckOfCertificates);
                offset += limit;
                moreRecordsToFetch = chuckOfCertificates.Count == limit;
                iteration++;
                
            }
            

            return giftCertificates;
        }
    }
}